x/wasm/IBC.md

```diff
@@ -0,0 +1,128 @@
+* `x/wasm` only accepts *ORDERED Channels* for simplicity of contract
+  correctness.
+* When sending a packet, the CosmWasm contract must specify the *ChannelID*
+  (and remote *PortID*?), which is generally provided by the external client
+  which understands which channels it wishes to communicate over.
+* When sending a packet, the contract can set a custom identifier (of
+  max 64 characters) that it can use to look this up later.
+* When receiving a Packet (or Ack or Error), the contracts receives the
+  *ChannelID* as well as remote *PortID* (and *ClientID*???) that is
+  communicating.
+* When receiving an Ack or Error packet, the contract also receives the
+  same identifier that it set on Send (`x/wasm` handles mapping between this
+  and the native IBC packet IDs).
+  
+### One Port per Module
+
+In this approach, the `x/wasm` module just binds one port to handle all
+modules. This can be well defined name like `wasm`. Since we always
+have `(ChannelID, PortID)` for routing messages, we can reuse one port
+for all contracts as long as we have a clear way to map the `ChannelID`
+to a specific contract when it is being established.
+
+
+* On genesis we bind the port `wasm` for all communication with the `x/wasm`
+  module.
+* The *Port* is fully owned by `x/wasm`
+* Each *Channel* is fully owned by one contract.
+* `x/wasm` only accepts *ORDERED Channels* for simplicity of contract
+  correctness.
+
+To clarify:
+
+* When a *Channel* is being established with port `wasm`, the
+  `x/wasm.Keeper` must be able to identify for which contract this
+  is destined. **how to do so**??
+  * One idea: the channel name must be the contract address. This means
+    (`wasm`, `cosmos13d...`) will map to the given contract in the wasm module.
+    The problem with this is that if two contracts from chainA want to
+    connect to the same contracts on chainB, they will want to claim the
+    same *ChannelID* and *PortID*. Not sure how to differentiate multiple
+    parties in this way.
+  * Other ideas: have a special field we send on `OnChanOpenInit` that
+    specifies the destination contract, and allow any *ChannelID*.
+    However, looking at [`OnChanOpenInit` function signature](https://docs.cosmos.network/master/ibc/custom.html#implement-ibcmodule-interface-and-callbacks),
+    I don't see a place to put this extra info, without abusing the version field,
+    which is a [specified field](https://docs.cosmos.network/master/ibc/custom.html#channel-handshake-version-negotiation):
+    ```
+    Versions must be strings but can implement any versioning structure. 
+    If your application plans to have linear releases then semantic versioning is recommended.
+    ... 
+    Valid version selection includes selecting a compatible version identifier with a subset 
+    of features supported by your application for that version.
+    ...    
+    ICS20 currently implements basic string matching with a
+    single supported version.
+    ```
```

Comment:
We could also put the sender/ receiver contract addresses as fields in the IBC packets we pass.